lib/sparesults/src/json.rs

```diff
@@ -224,520 +220,882 @@ fn write_json_term<'a>(output: &mut Vec<JsonEvent<'a>>, term: TermRef<'a>) {
+                            level: 0,
+                            after: JsonInnerReaderStateAfterIgnore::InRootObject,
+                        };
+                        Ok(None)
+                    }
+                },
+                JsonEvent::EndObject => Err(QueryResultsSyntaxError::msg(
+                    "SPARQL JSON results must contain a 'boolean' or a 'results' key",
+                )),
+                _ => unreachable!(),
+            },
+            JsonInnerReaderState::BeforeHead => {
+                if event == JsonEvent::StartObject {
+                    self.state = JsonInnerReaderState::InHead;
+                    Ok(None)
+                } else {
+                    Err(QueryResultsSyntaxError::msg(
+                        "SPARQL JSON results head must be an object",
+                    ))
+                }
+            }
+            JsonInnerReaderState::InHead => match event {
+                JsonEvent::ObjectKey(key) => match key.as_ref() {
+                    "vars" => {
+                        self.state = JsonInnerReaderState::BeforeVars;
+                        self.vars_read = true;
+                        Ok(None)
+                    }
+                    "links" => {
+                        self.state = JsonInnerReaderState::BeforeLinks;
+                        Ok(None)
+                    }
+                    _ => {
+                        self.state = JsonInnerReaderState::Ignore {
+                            level: 0,
+                            after: JsonInnerReaderStateAfterIgnore::InHead,
+                        };
+                        Ok(None)
+                    }
+                },
+                JsonEvent::EndObject => {
+                    self.state = JsonInnerReaderState::InRootObject;
+                    Ok(None)
+                }
+                _ => unreachable!(),
+            },
+            JsonInnerReaderState::BeforeVars => {
+                if event == JsonEvent::StartArray {
+                    self.state = JsonInnerReaderState::InVars;
+                    Ok(None)
+                } else {
+                    Err(QueryResultsSyntaxError::msg(
+                        "SPARQL JSON results vars must be an array",
+                    ))
+                }
+            }
+            JsonInnerReaderState::InVars => match event {
+                JsonEvent::String(variable) => match Variable::new(variable.clone()) {
+                    Ok(var) => {
+                        if self.variables.contains(&var) {
+                            return Err(QueryResultsSyntaxError::msg(format!(
+                                "The variable {var} is declared twice"
+                            )));
+                        }
+                        self.variables.push(var);
+                        Ok(None)
                     }
+                    Err(e) => Err(QueryResultsSyntaxError::msg(format!(
```

Comment:
Usually it is advised not to use formatting to create an error message because later that message might actually be dropped without being used. Instead, there should be an error enum variant with all the needed values - so that an error message can be constructed at the last moment if needed.

Reply:
Indeed. Thank you! But I am not sure it's very significant, we already allocate quite a lot when building the parser and returning the results (each `Term` wraps at least an allocation) so an extra allocation on error is probably not a big deal. I would tend to think there are a lot of things to optimize in this parser and elsewhere is Oxigraph.

Reply:
Its probably not in this case...  To me, this is more of a "good practice to always do" -- this way you never even have to think if this is a perf bottleneck or not, and keeps the code more compact, and allows to track all possible error conditions:

```rust
Err(QueryResultsSyntaxError::msg(
                    "SPARQL JSON results must contain a 'boolean' or a 'results' key",
                )),

// vs

Err(QueryResultsSyntaxError::JsonMissingBoolOrResultsKey),
```

Reply:
Yes! Agreed! Just got lazy with this MR. It's a pervasise thing in Oxigraph codebase. I have opened #790 about it.